### PR TITLE
Added Select

### PR DESCRIPTION
**Pull request by CarloPalinckx**
### This PR:

- [x] needs tests
- [x] need to look at the open animation, placeholder has some weird movement when opening.

**Adds** ✨
- A filterable `Select` component
- A `maxHeight` prop to `Box`

**Changes** 🌀
- Inline behavior for `Box` it now renders a `span` when `inline` is set to true.
- Fixes a bug where `autoHideScrollBar` on `ScrollBox` would be ignored.
- Confines `Contrast`'s theme merging behavior to construction instead of possibly every render. This should increase its performance.

**Removes** 👋
- Tslint ordered import rules. If we really care deeply about import ordening we can run the tslint fixer to sort them for us occasionally.
- `Box`'s test snapshot. These were way too brittle and now just assert the style rule applied.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -86,6 +85,7 @@
     "lint-fix": "tslint --fix --project ./config/typescript/tsconfig.json --config ./config/typescript/tslint.json ./src/**/*.{ts,tsx}",
     "lint": "tslint --project ./config/typescript/tsconfig.json --config ./config/typescript/tslint.json ./src/**/*.{ts,tsx}",
     "watch": "start-storybook -p 9001 -c ./config/storybook",
+    "dev": "npm run watch",
```
Comment (woutsluiter) [inline package.json]: Why is this added?

File: package.json [hunk 2]
```diff
@@ -19,7 +19,6 @@
     "@types/react": "^16.3.11",
     "@types/react-dom": "^16.0.5",
     "deepmerge": "^2.1.0",
-    "react-popper": "^1.0.0-beta.6",
     "resize-observer-polyfill": "^1.5.0",
```
[the remark this answers is not in the document]
Comment (CarloPalinckx) [reply package.json]: Something probably went wrong with rebasing, ill revert this 👍

File: package.json (hunk 1 above)
Comment (CarloPalinckx) [reply package.json]: just an alias, because the conventional name for starting a dev server is `dev`

**Comment by CarloPalinckx** [comment]
> when you set a predefined/selected value, the idle state shows the correct label ("Bar F"), but when you open the Select it shows the value ("F"). I would expect it to be empty since I didn't type anything yet.

could you shoot a gif of this? Can't seem to reproduce it.

> The results are always one cycle behind the typed value
used the state input value before updating it, its using event value directly now. 😅

File: src/components/Select/style.tsx [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+type WindowProps = {
+    isOpen: boolean;
+};
+
+const StyledWindow = withProps<WindowProps, HTMLDivElement>(styled.div)`
+    transition: border .3s, width .3s, left .3s;
+    position: absolute;
+    max-height: 240px;
+    overflow: hidden;
+    width: ${({ isOpen }): string => (isOpen ? '100%' : 'calc(100% - 4px)')};
```
Comment (woutsluiter) [inline src/components/Select/style.tsx]: The whole search field seems to resize 1px. This makes the whole transition a bit jerky. I think it would feel a lot smoother if the search field wouldn't resize/move at all.

**Comment by woutsluiter** [comment]
Lol FF
image

**Comment by CarloPalinckx** [comment]
The icon is just shy in a new environment 😄

**Comment by woutsluiter** [comment]
>could you shoot a gif of this? Can't seem to reproduce it.
It occurs when you set the `state.value` in the story to "F" for example.

foo

File: src/components/Select/index.tsx [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+            this.state.optionPointer < this.state.filteredOption.length ? this.state.optionPointer + 1 : 0;
+
+        this.setState({ optionPointer });
+    };
+
+    private cycleDown = (): void => {
+        const optionPointer =
+            this.state.optionPointer > 0 ? this.state.optionPointer - 1 : this.state.filteredOption.length;
+
+        this.setState({ optionPointer });
```
Comment (woutsluiter) [inline src/components/Select/index.tsx]: I'm getting these TS errors:
`Type argument cannot be inferred from usage`

File: src/components/Select/index.tsx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    public componentDidMount(): void {
+        document.addEventListener('mousedown', this.handleClickOutside);
+    }
+
+    public componentWillUnmount(): void {
+        document.removeEventListener('mousedown', this.handleClickOutside);
+    }
+
+    public handleClickOutside = (event: MouseEvent): void => {
+        if (this.wrapperRef !== null && !this.wrapperRef.contains(event.target as Node)) {
```
Comment (woutsluiter) [inline src/components/Select/index.tsx]: `Binary operation argument type HTMLDivElement is not compatible with type null`
